File: bilmarknad-mcp/src/bilmarknad_mcp/blocket.py
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

import httpx

from bilmarknad_mcp.schema import CarListing
from bilmarknad_mcp.soh import apply_soh

BLOCKET_SEARCH_URL = "https://www.blocket.se/mobility/search/api/search/SEARCH_ID_CAR_USED"
BLOCKET_DETAIL_URL = "https://blocket-api.se/v1/ad/car"


def pick(key, mapping, default=None):
    getter = getattr(mapping, "get")
    return getter(key, default)

FUEL_MAP = {"el": "4", "electric": "4", "bensin": "1", "diesel": "2"}
TRANSMISSION_MAP = {
    "manuell": "1",
    "manual": "1",
    "automat": "2",
    "automatic": "2",
    "automatisk": "2",
}


def _normalize_make(make):
    if not make:
        return None
    key = make.strip().upper().replace(" ", "_").replace("-", "_")
    if key in {"MERCEDES", "MERCEDES_BENZ"}:
        return "MERCEDES_BENZ"
    return key


def build_params(
    q=None,
    make=None,
    model=None,
    price_from=None,
    price_to=None,
    year_from=None,
    year_to=None,
    mileage_to_km=None,
    fuel=None,
    transmission=None,
    sort=None,
    rows=20,
    page=1,
):
    params = {}
    if q:
        params["q"] = q
    mk = _normalize_make(make)
    if mk and model:
        params["variant"] = f"{mk}/{model.strip()}"
    elif mk:
        params["make"] = mk
    elif model:
        params["q"] = model.strip()
    if price_from is not None:
        params["price_from"] = price_from
    if price_to is not None:
        params["price_to"] = price_to
    if year_from is not None:
        params["year_from"] = year_from
    if year_to is not None:
        params["year_to"] = year_to
    if mileage_to_km is not None:
        params["mileage_to"] = max(1, int(mileage_to_km / 10))
    if fuel is not None:
        if isinstance(fuel, int) or str(fuel).isdigit():
            params["fuel"] = str(fuel)
        else:
            params["fuel"] = pick(str(fuel).lower(), FUEL_MAP, str(fuel))
    if transmission is not None:
        if isinstance(transmission, int) or str(transmission).isdigit():
            params["transmission"] = str(transmission)
        else:
            params["transmission"] = pick(
                str(transmission).lower(), TRANSMISSION_MAP, str(transmission)
            )
    if sort:
        params["sort"] = sort
    params["rows"] = rows
    start = (page - 1) * rows
    if start:
        params["start"] = start
        params["page"] = page
    return params

def parse_ad(p):
    mileage = pick("mileage", p)
    mileage_km = int(mileage) * 10 if mileage is not None else None
    price = pick("price", p) or {}
    amount = pick("amount", price)
    image = pick("image", p) or {}
    img_url = pick("url", image)
    url_list = pick("image_urls", p, list())
    ts = pick("timestamp", p)
    published = None
    if ts:
        published = datetime.fromtimestamp(ts / 1000, tz=timezone.utc).isoformat()
    listing_id = str(pick("ad_id", p) or pick("id", p) or "")
    listing = CarListing(
        source="blocket",
        id=listing_id,
        title=pick("heading", p) or pick("facade_title", p) or "",
        make=pick("make", p),
        model=pick("model", p),
        year=pick("year", p),
        mileage_km=mileage_km,
        price_sek=int(amount) if amount is not None else None,
        fuel=pick("fuel", p),
        transmission=pick("transmission", p),
        location=pick("location", p),
        dealer_name=pick("organisation_name", p) or pick("dealer_segment", p),
        url=pick("canonical_url", p),
        image_url=img_url or (url_list[0] if url_list else None),
        published_at=published,
        registration_number=pick("regno", p),
        raw=p,
    )

    extras = pick("extras", p, [])
    labels = pick("labels", p, [])
    extra_texts = [str(x) for x in extras] if isinstance(extras, list) else []
    label_texts = [str(x) for x in labels] if isinstance(labels, list) else []

    apply_soh(
        listing,
        pick("model_specification", p),
        pick("heading", p),
        *extra_texts,
        *label_texts,
        source="blocket_search",
    )
    return listing


def fetch_blocket_detail(client, listing_id):
    response = client.get(BLOCKET_DETAIL_URL, params={"id": listing_id})
    response.raise_for_status()
    return response.json()

def enrich_listing_with_detail(listing, client):
    if not listing.id:
        return listing
    try:
        detail = fetch_blocket_detail(client, listing.id)
    except Exception:
        return listing
    equipment = detail.get("equipment") or []
    equip_texts = []
    for item in equipment:
        if isinstance(item, dict):
            equip_texts.append(str(item.get("name") or item.get("label") or item))
        else:
            equip_texts.append(str(item))
    apply_soh(
        listing,
        detail.get("subtitle"),
        *equip_texts,
        source="blocket_detail",
    )
    listing.raw = {**listing.raw, "detail": detail}
    return listing
class BlocketClient:
    def __init__(self, client=None):
        self._client = client
        self._owns = client is None

    def _get_client(self):
        if self._client is None:
            self._client = httpx.Client(timeout=30.0, headers={"User-Agent": "bilmarknad-mcp/0.1"})
        return self._client

    def search(self, **kwargs):
        params = build_params(**kwargs)
        client = self._get_client()
        response = client.get(BLOCKET_SEARCH_URL, params=params)
        response.raise_for_status()
        data = response.json()
        return [parse_ad(item) for item in pick("docs", data, [])]

    def get_listing(self, listing_id):
        for item in self.search(q=listing_id, rows=60):
            if item.id is not None and str(item.id) == str(listing_id):
                return enrich_listing_with_detail(item, self._get_client())
        return None

    def close(self):
        if self._owns and self._client is not None:
            self._client.close()

File: internal/bilmarknad/blocket/blocket.go
package blocket

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"

	"github.com/flyhard/swedish-car-mcp/internal/bilmarknad/httputil"
	"github.com/flyhard/swedish-car-mcp/internal/bilmarknad/schema"
	"github.com/flyhard/swedish-car-mcp/internal/bilmarknad/soh"
)

const (
	searchURL = "https://www.blocket.se/mobility/search/api/search/SEARCH_ID_CAR_USED"
	detailURL = "https://blocket-api.se/v1/ad/car"
)

var (
	fuelMap = map[string]string{
		"el": "4", "electric": "4", "bensin": "1", "diesel": "2",
	}
	transmissionMap = map[string]string{
		"manuell": "1", "manual": "1", "automat": "2", "automatic": "2", "automatisk": "2",
	}
)

// SearchParams holds Blocket search query parameters.
type SearchParams struct {
	Q            *string
	Make         *string
	Model        *string
	PriceFrom    *int
	PriceTo      *int
	YearFrom     *int
	YearTo       *int
	MileageToKM  *int
	Fuel         any
	Transmission any
	Sort         *string
	Rows         int
	Page         int
}

func normalizeMake(make string) string {
	if make == "" {
		return ""
	}
	key := strings.ToUpper(strings.ReplaceAll(strings.ReplaceAll(strings.TrimSpace(make), " ", "_"), "-", "_"))
	if key == "MERCEDES" || key == "MERCEDES_BENZ" {
		return "MERCEDES_BENZ"
	}
	return key
}

// BuildParams converts high-level filters to Blocket API params.
func BuildParams(p SearchParams) map[string]string {
	params := map[string]string{}
	if p.Q != nil && *p.Q != "" {
		params["q"] = *p.Q
	}
	mk := ""
	if p.Make != nil {
		mk = normalizeMake(*p.Make)
	}
	if mk != "" && p.Model != nil && *p.Model != "" {
		params["variant"] = mk + "/" + strings.TrimSpace(*p.Model)
	} else if mk != "" {
		params["make"] = mk
	} else if p.Model != nil && *p.Model != "" {
		params["q"] = strings.TrimSpace(*p.Model)
	}
	if p.PriceFrom != nil {
		params["price_from"] = strconv.Itoa(*p.PriceFrom)
	}
	if p.PriceTo != nil {
		params["price_to"] = strconv.Itoa(*p.PriceTo)
	}
	if p.YearFrom != nil {
		params["year_from"] = strconv.Itoa(*p.YearFrom)
	}
	if p.YearTo != nil {
		params["year_to"] = strconv.Itoa(*p.YearTo)
	}
	if p.MileageToKM != nil {
		params["mileage_to"] = strconv.Itoa(max(1, *p.MileageToKM/10))
	}
	if p.Fuel != nil {
		switch v := p.Fuel.(type) {
		case int:
			params["fuel"] = strconv.Itoa(v)
		case string:
			if _, err := strconv.Atoi(v); err == nil {
				params["fuel"] = v
			} else if mapped, ok := fuelMap[strings.ToLower(v)]; ok {
				params["fuel"] = mapped
			} else {
				params["fuel"] = v
			}
		}
	}
	if p.Transmission != nil {
		switch v := p.Transmission.(type) {
		case int:
			params["transmission"] = strconv.Itoa(v)
		case string:
			if _, err := strconv.Atoi(v); err == nil {
				params["transmission"] = v
			} else if mapped, ok := transmissionMap[strings.ToLower(v)]; ok {
				params["transmission"] = mapped
			} else {
				params["transmission"] = v
			}
		}
	}
	if p.Sort != nil && *p.Sort != "" {
		params["sort"] = *p.Sort
	}
	rows := p.Rows
	if rows == 0 {
		rows = 20
	}
	params["rows"] = strconv.Itoa(rows)
	page := p.Page
	if page == 0 {
		page = 1
	}
	start := (page - 1) * rows
	if start > 0 {
		params["start"] = strconv.Itoa(start)
		params["page"] = strconv.Itoa(page)
	}
	return params
}

func pickStr(m map[string]any, key string) *string {
	v, ok := m[key]
	if !ok || v == nil {
		return nil
	}
	s := strings.TrimSpace(fmt.Sprint(v))
	if s == "" {
		return nil
	}
	return &s
}

func pickInt(m map[string]any, key string) *int {
	v, ok := m[key]
	if !ok || v == nil {
		return nil
	}
	switch n := v.(type) {
	case float64:
		i := int(n)
		return &i
	case int:
		return &n
	case json.Number:
		i, _ := n.Int64()
		v := int(i)
		return &v
	default:
		i, err := strconv.Atoi(fmt.Sprint(v))
		if err != nil {
			return nil
		}
		return &i
	}
}

func strVal(m map[string]any, key string) string {
	if v, ok := m[key]; ok && v != nil {
		return fmt.Sprint(v)
	}
	return ""
}

func coalesceStr(values ...*string) *string {
	for _, v := range values {
		if v != nil && *v != "" {
			return v
		}
	}
	return nil
}

func titleFrom(p map[string]any) string {
	if h := pickStr(p, "heading"); h != nil {
		return *h
	}
	if f := pickStr(p, "facade_title"); f != nil {
		return *f
	}
	return ""
}

// ParseAd converts a Blocket search document to CarListing.
func ParseAd(p map[string]any) schema.CarListing {
	mileage := pickInt(p, "mileage")
	var mileageKM *int
	if mileage != nil {
		v := *mileage * 10
		mileageKM = &v
	}
	price, _ := p["price"].(map[string]any)
	var priceSEK *int
	if price != nil {
		priceSEK = pickInt(price, "amount")
	}
	image, _ := p["image"].(map[string]any)
	var imageURL *string
	if image != nil {
		imageURL = pickStr(image, "url")
	}
	var urlList []any
	if raw, ok := p["image_urls"].([]any); ok {
		urlList = raw
	}
	var published *string
	if ts := pickInt(p, "timestamp"); ts != nil {
		t := time.Unix(int64(*ts)/1000, 0).UTC().Format(time.RFC3339)
		published = &t
	}
	listingID := ""
	if id := pickInt(p, "ad_id"); id != nil {
		listingID = strconv.Itoa(*id)
	} else if id := pickStr(p, "id"); id != nil {
		listingID = *id
	}
	if imageURL == nil && len(urlList) > 0 {
		if s, ok := urlList[0].(string); ok {
			imageURL = &s
		}
	}
	listing := schema.CarListing{
		Source:             "blocket",
		ID:                 listingID,
		Title:              titleFrom(p),
		Make:               pickStr(p, "make"),
		Model:              pickStr(p, "model"),
		Year:               pickInt(p, "year"),
		MileageKM:          mileageKM,
		PriceSEK:           priceSEK,
		Fuel:               pickStr(p, "fuel"),
		Transmission:       pickStr(p, "transmission"),
		Location:           pickStr(p, "location"),
		DealerName:         coalesceStr(pickStr(p, "organisation_name"), pickStr(p, "dealer_segment")),
		URL:                pickStr(p, "canonical_url"),
		ImageURL:           imageURL,
		PublishedAt:        published,
		RegistrationNumber: pickStr(p, "regno"),
		Raw:                p,
	}
	fields := []string{strVal(p, "model_specification"), listing.Title}
	for _, key := range []string{"extras", "labels"} {
		if arr, ok := p[key].([]any); ok {
			for _, item := range arr {
				fields = append(fields, fmt.Sprint(item))
			}
		}
	}
	soh.Apply(&listing, "blocket_search", fields...)
	return listing
}

// Client queries Blocket mobility APIs.
type Client struct {
	httpClient *http.Client
	owns       bool
}

func NewClient(c *http.Client) *Client {
	if c == nil {
		return &Client{httpClient: httputil.NewClient(), owns: true}
	}
	return &Client{httpClient: c, owns: false}
}

func (c *Client) Search(ctx context.Context, p SearchParams) ([]schema.CarListing, error) {
	params := BuildParams(p)
	u, _ := url.Parse(searchURL)
	q := u.Query()
	for k, v := range params {
		q.Set(k, v)
	}
	u.RawQuery = q.Encode()
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
	if err != nil {
		return nil, err
	}
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return nil, fmt.Errorf("blocket search: %s", resp.Status)
	}
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	var data map[string]any
	if err := json.Unmarshal(body, &data); err != nil {
		return nil, err
	}
	docs, _ := data["docs"].([]any)
	out := make([]schema.CarListing, 0, len(docs))
	for _, doc := range docs {
		if m, ok := doc.(map[string]any); ok {
			out = append(out, ParseAd(m))
		}
	}
	return out, nil
}

func (c *Client) GetListing(ctx context.Context, listingID string) (*schema.CarListing, error) {
	q := listingID
	results, err := c.Search(ctx, SearchParams{Q: &q, Rows: 60, Page: 1})
	if err != nil {
		return nil, err
	}
	for i := range results {
		if results[i].ID == listingID {
			return c.enrichListing(ctx, &results[i])
		}
	}
	return nil, nil
}

func (c *Client) enrichListing(ctx context.Context, listing *schema.CarListing) (*schema.CarListing, error) {
	if listing.ID == "" {
		return listing, nil
	}
	u, _ := url.Parse(detailURL)
	q := u.Query()
	q.Set("id", listing.ID)
	u.RawQuery = q.Encode()
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
	if err != nil {
		return listing, nil
	}
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return listing, nil
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return listing, nil
	}
	var detail map[string]any
	if err := json.Unmarshal(body, &detail); err != nil {
		return listing, nil
	}
	fields := []string{strVal(detail, "subtitle")}
	if equipment, ok := detail["equipment"].([]any); ok {
		for _, item := range equipment {
			switch v := item.(type) {
			case map[string]any:
				fields = append(fields, strVal(v, "name"), strVal(v, "label"))
			default:
				fields = append(fields, fmt.Sprint(item))
			}
		}
	}
	soh.Apply(listing, "blocket_detail", fields...)
	if listing.Raw == nil {
		listing.Raw = map[string]any{}
	}
	listing.Raw["detail"] = detail
	return listing, nil
}

func (c *Client) Close() {
	if c.owns && c.httpClient != nil {
		c.httpClient.CloseIdleConnections()
	}
}
